### ISSUE TITLE
FAIL: t2800-jobs-cmd.t 61 - flux-jobs --format={waitstatus},{returncode}

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#5210]
during large parallel make checks I've recently gotten this error twice, the error comes from the fact that the `returncodeI.out` output is:

```
-,-128
36352,142
32512,127
36608,143
32512,127
0,0
0,0
0,0
0,0
```

but expected is

```
-,-128
36352,142
36608,143
36608,143
32512,127
0,0
0,0
0,0
0,0
```

The invalid `32512,127` says that the submitted command was not found, instead of the command being terminated via an exception (which is `36608,143`).  The failing job is:

```
        #
        #  Run a job that we will end with a user exception, copy its JOBID to both
        #       inactive and failed and exception lists.
        #
        jobid=`flux submit --wait-event=start sleep inf` &&
        flux job raise --type=myexception --severity=0 -m "myexception" $jobid &&
        fj_wait_event $jobid clean &&
        echo $jobid >> inactiveids &&
        flux job id $jobid > exception.ids &&
        flux job id $jobid >> failedids &&
```

I presently don't see how something busy / racy could make this fail with "command not found".

Will just monitor this for the time being.

**Comment by chu11** [comment]
I've been carrying around a stupid debug patch to try and get debug info b/c I've seen this ehh ... 4-5 times total since I first hit this.  It's quite rare.  Hit it today with some debug info.  So here's the eventlog for the bad job in question.

```
1687971966.198527 submit userid=8556 urgency=16 flags=0 version=1
1687971966.210447 validate
1687971966.221623 depend
1687971966.221652 priority priority=16
1687971966.223060 alloc annotations={"sched":{"resource_summary":"rank0/core0"}}
1687971966.227694 start
1687971966.310803 exception type="myexception" severity=0 userid=8556 note="myexception"
1687971966.312935 exception type="exec" severity=0 userid=8556 note="task 0 (host corona212): start failed: sleep: No such file or directory"
1687971966.315559 finish status=32512
1687971966.317775 release ranks="all" final=true
1687971966.318339 free
1687971966.318353 clean
```

The most notable thing is that there are TWO exception events.  The one that I generate myself (the "myexception" one) and the one generated by the shell ("exec").  The "start failed" error means an error occurred in `shell_task_start()` (which most likely means an error occurs in `flux_local_exec_ex()`), and leads to `shell_task_start()` to fail and generate that exception.  The wait status in the "finish" event "32512" maps to the second exception's "No such file or directory" / ENOENT.

So that much makes sense.  It does not appear `job-list` has an error.

But it's still a mystery to me as to why the second exception is happening and how ENOENT occurs.

The first "myexception" leads to the job shell being sent a SIGTERM from `job-exec`.

- The job shell wants to forward that signal to the task, but forwarding the signal should be impossible b/c there are no tasks running (we know this b/c `shell_task_start()` fails and this is a one task job).

  SIDE QUESTION: Is this a corner case?  Signal received by shell before any tasks running.  Should shell exit?  It appears it would do nothing right now if there are no tasks running.  So tasks run after this signal would still get started?  Or perhaps it doesn't matter, a later SIGKILL from job-exec would kill everything.

  Soooo ... how does `shell_task_start()` fail and why is the error ENOENT?  I'm not entirely sure at the moment.  The timestamps of the eventlog above indicate that something happens very quickly, so this is not the SIGKILL coming at a later time.  I'm wondering if something else triggers a shutdown in the shell that I'm not seeing?  The pre-exec callback does not seem to do anything special in this case.

- ALTERNATE THOUGHT: Internally `flux_local_exec_ex()` may have already done a `fork()`.  Is it possible SIGTERM gets forwarded to the task?  We set `FLUX_SUBPROCESS_FLAGS_SETPGRP` by default, and the subprocess does not trap SIGTERM.  I'm not sure if only trapping the signal in the parent of the process in a process group is enough?  Even if this is true, I'm not sure why `flux_local_exec_ex()` would return ENOENT.  Possibly a lingering errno in the environment when it returned is my only current guess.

- The other question of course is if this can be handled?  In some way in `job-list`?  `job-exec` and `job-list` are acting accordingly at the moment.  The main issue is the shell exiting in this unexpected way.

**Comment by grondo** [comment]
If I had to guess, I'd also presume that `ENOENT` is a stale `errno` left from some other call.

>  Is this a corner case? Signal received by shell before any tasks running.

Yeah, this sounds like an oversight. Maybe the shell should block signals until all tasks have been started, then presumably once the signal watcher is started, the signal would be handled and tasks terminated appropriately. Another, perhaps simpler solution would be to have the shell exit with the appropriate exit code when no tasks have started yet and a signal is delivered.

That small fix may avoid the second exception, and we don't have to worry about where the `ENOENT` is coming from?

**Comment by chu11** [comment]
> That small fix may avoid the second exception, and we don't have to worry about where the ENOENT is coming from?

```
void flux_shell_killall (flux_shell_t *shell, int signum)                                                                                   
{                                                                                                                                           
    struct shell_task *task;                                                                                                                
    if (!shell ||  signum <= 0 || !shell->tasks)                                                                                            
        return;                                                                                                                             
    task = zlist_first (shell->tasks);                                                                                                      
    while (task) {                                                                                                                          
        if (shell_task_running (task) && shell_task_kill (task, signum) < 0)                                                                
            shell_log_errno ("kill task %d: signal %d", task->rank, signum);                                                                
        task = zlist_next (shell->tasks);                                                                                                   
    }                                                                                                                                       
} 
```

Yeah, I look at something like this and think "if there are no tasks ... we should just exit?", i.e. just "send" the signal to ourselves.

Although I'm still perplexed how `flux_local_exec_ex()` is returning an error.

**Comment by grondo** [comment]
Should probably only kill self if the signal is fatal. I.e. just ignore `SIGUSR1` and `SIGUSR2`?

**Comment by chu11** [comment]
> Should probably only kill self if the signal is fatal. I.e. just ignore SIGUSR1 and SIGUSR2?

Interesting point.  I would have originally thought to kill under all signals since that is default behavior, but maybe not.  Perhaps should consider your other idea of forwarding signals after all tasks have started.

**Comment by grondo** [comment]
Well, my thinking is that the user intends to send SIGUSR1 to cause some application behavior, e.g. a restart dump. It might be better to ignore the signal in this case rather than terminating the job before it starts. However, this case is pretty unlikely since a user won't be sending this kind of control signal so early, so I'm fine with terminating the job shell for now since that's much simpler, and we'll see if we get any bug reports.

**Comment by chu11** [comment]
> However, this case is pretty unlikely since a user won't be sending this kind of control signal so early, so I'm fine with terminating the job shell for now since that's much simpler, and we'll see if we get any bug reports.

I had started to try the "wait until tasks have started and then forward signal", but as I thought about it, there's also the race that the job tasks have to setup their signal handlers too.  No guarantees that they have set things up by the time we forward the signal, especially so quickly after the launch.

As you said, this is such a unique and small racy corner case, perhaps just terminate or ignore a few signals is fine.  Documenting the potential race in the code is probably more important.

**Comment by chu11** [comment]
Implemented a fix and by dumb luck hit the "FAIL: t2800-jobs-cmd.t 61 - flux-jobs --format={waitstatus},{returncode}" error again.  Always a chance I implemented the fix wrong, but perhaps not related to the original issue.  (Although I think the fix is a legit fix, as it is a potential corner case).

Tried to reproduce with some scripts looping through 

```
        jobid=`flux submit --wait-event=start sleep inf` &&
        flux job raise --type=myexception --severity=0 -m "myexception" $jobid &&
```

over and over.  Just can't get that perfect mix of busy that is causing this.

**Comment by chu11** [comment]
one other note, now that I have a clue what is causing this, we could solve this issue via something like:

```
> cat foo.sh
#!/bin/sh
echo "started my job"
sleep inf
```

```
jobid=`flux submit --wait-event=start foo.sh` &&
while ! flux job attach $jobid | grep "started my job"
do
     sleep 1
done
flux job raise --type=myexception --severity=0 -m "myexception" $jobid
```

but not 100% sure want to do this?  It's quite rare and would be good to eventually figure out root cause.

**Comment by grondo** [comment]
No, we want to avoid `sleep 1` in tests as much as possible. 
However, wouldn't this work to avoid the kludgy while loop?
```
flux job wait-event -v -p guest.output $jobid data
``` 
i.e. wait for the first `data` event to hit the output eventlog.

**Comment by chu11** [comment]
>  However, wouldn't this work to avoid the kludgy while loop?

Ahhh that's a good idea to wait-event on the output stream.  I think that would work.

**Comment by chu11** [comment]
I had an idea this morning on how to sorta-reproduce this issue, although I'm not going to make a test since it's still racy.  So mostly just putting here for documentation.

```
#!/bin/sh

test_description='debug'

. $(dirname $0)/job-list/job-list-helper.sh

. $(dirname $0)/sharness.sh

export TASKCOUNT=1000
export TEST_UNDER_FLUX_CORES_PER_RANK=${TASKCOUNT}
test_under_flux 1 job

test_expect_success 'test' '
                    jobid=`flux submit --wait-event=start -N1 -n${TASKCOUNT} sleep inf` &&
                    flux job raise --type=myexception --severity=0 -m "myexception" $jobid &&
                    flux job eventlog $jobid &&
                    ! flux job attach -E -X -v $jobid &&
                    flux job wait-event --timeout=20 $jobid clean
'

test_expect_success 'flux jobs' '
                    flux jobs -a -o "{id},{ntasks},{waitstatus},{returncode}"
'

test_done
```

Basically with a high task count we can increase the odds the tasks haven't completed setting up when we raise the exception.  Had to tweak the task count a bit here and there, temporarily remove the shell's affinity plugin, and sometimes I would mess with putting some sleeps in between the job submission and raising the exception.  Basically it's still racy, I figure high task count initialization and perhaps barrier stuff makes it a little racy still.

In the master branch, with this technique I often get

```
JOBID,NTASKS,WSTATUS,RC
f2SFefr7,1000,15,-15
```

(15 = SIGTERM)

which I assume the signal handler hasn't been setup yet, and actually saw this a few times ...

```
JOBID,NTASKS,WSTATUS,RC
f2PoKro1,1000,9,-9
```

(9 = SIGKILL)

I assume the signal caught after the original trap set AND < 1000 tasks setup so shell didn't exit, so it is eventually SIGKILL-ed.

In the branch with fixes in #5299 still get 15/-15 sometimes, but do see:

```
JOBID,NTASKS,WSTATUS,RC
f2NUEW95,1000,36608,143
```

and

```
0.143s: flux-shell[0]: FATAL: signals: signals.c:35: signal 15 sent before tasks started
```

(FWIW, a `sleep 0.25` between the job submit and raise seems to be the sweet spot for me to getting exit 143 through the "early signal caught".)

So this made me realize that outside of extreme measures, this is just inherently racy part of job launch.  I think the fix in #5299 is "good" and "makes things better", but is certainly not perfect.

The solution discussed above should be done to ensure the test in `t2800-jobs-cmd.t` should always succeed.

**Comment by chu11** [comment]
Closed accidentally bc of an errant "fixes ..." Comment in #5289.  Reopening
